x/gov/keeper/msg_server.go

```diff
@@ -78,6 +83,32 @@ func (k msgServer) SubmitProposal(goCtx context.Context, msg *v1.MsgSubmitPropos
 	}, nil
 }
 
+func (k msgServer) CancelProposal(goCtx context.Context, msg *v1.MsgCancelProposal) (*v1.MsgCancelProposalResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	_, err := sdk.AccAddressFromBech32(msg.Proposer)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
If we do the above change (i.e. no migrations at all), then we also need to add this:

```suggestion
	if msg.Proposer == "" {
		// Backwards compatibility: an empty proposer means that
		// the proposals is an old one (i.e. submitted before the MsgCancelProposal
		// feature was introduced. In this case, we don't allow canceling.
		return nil, fmt.Errorf("proposal %d was submitted before the upgrade, so cannot be canceled", msg.ProposalId)
	}
```

Reply:
@anilCSE  what do you think should we keep cancel proposal only to new proposal ( after upgrade)

Reply:
Yes, i vote for that

Reply:
I think we should keep migrations 

```go
// CancelProposal will cancel proposal before the voting period ends
func (keeper Keeper) CancelProposal(ctx sdk.Context, proposalID uint64, proposer string) error {
	proposal, ok := keeper.GetProposal(ctx, proposalID)
	if !ok {
		return sdkerrors.Wrapf(types.ErrProposalNotFound, "proposal not found with %d id", proposalID)
	}

	if proposal.Proposer == "" {
		return sdkerrors.Wrapf(types.ErrInvalidProposal, "proposal %d does n't have proposer %s", proposalID, proposer)
	}
	
	....
}
``` 
with above conditions we can  allow `CancelProposal ` to which `proposal` will have `proposer`. 


1. If migrations done then proposals will have proposer to each one.
2. If proposal not have proposer (means old proposal) , we did not allow `CancelProposal`.
3. Anyway new proposal will have proposer

Reply:
Okay I will keep proposal migration function, proposal migration will be up to specific chain developers only and I will add conditions in migration to make sure migration only for current proposals (active proposals) not old proposals.

Reply:
@alexanderbez  Please can you review again

Reply:
> Then, the question remains, do we add Proposer to current proposals, which would need migrations. My argument is that the migration complexity is not worth it for just a few current proposals. Hence, my suggestion of removing the entire migration.

The argument for doing a trivial migration (setting `Proposer = ""`) is to make a consistent data layout. Will deserialization fail if trying to deserialize old record with the new structure?

Why do you think it's costly? We don't have thousands of proposals.

Reply:
Proposals are large and there is no reason to set a field to `""`. It will be empty by default in client responses.

Reply:
We are talking about tens of proposals (few hundreds in case of osmosis), with something like 1k of characters. It's not that much. But we can make it optional as already suggested.

Reply:
The point is that there is simply no need to do this. The proposer field will already be empty in those cases.